docs/hugo/content/design/ADR-2022-11-Resource-Import.md

```diff
@@ -0,0 +1,78 @@
+
+Since many of these resources are already supported by ASO v2, albeit with shapes much more closely aligned with the ARM originals, we would need to implement forward and backward conversions between the two shapes, preserving the Kubernetes version compatibility guarantees we have today.
+
+* Pro: Very simple migration for our customers.
+* Con: Would require a lot of work to implement and test the conversions.
+* Con: Any resource features that don't map exactly to a property in the ARM shape would be difficult to retain.
+* Con: The investment would benefit only a small number of customers.
+* Con: The conversions would need to be maintained, potentially indefinitely.
+* Con: It's unlikely we'd achieve 100% fidelity, requiring customers to change their YAML files anyway.
+
+### Resource Migration
+
+Create a tool that allowed any supported Azure resource to be to be exported as a YAML file in the shape expected by ASO v2. Users would then be able to modify the YAML file as necessary and apply it to their cluster.
+
+The most likely form of this tool would be as a command-line utility for users to run locally. Idiomatically for the Kubernetes ecosystem, we'd call this `asoctl`.
+
+* Pro: Straightforward migration for ASO v1 customers (though, not zero touch).
+* Pro: Would benefit all customers wanting to migrate existing Azure resources, not just those using ASO v1.
+* Pro: Could also be used to snapshot existing resources (e.g. to capture a hand configured resource as YAML for reuse).
+* Con: We'd need to extend the code generator to provide the required support (but much of what we need already exists).
+
+## Decision
+
+Create a new command line tool, `asoctl` to house multiple functions, starting with resource import. The tool will follow the usual convention of supporting a variety of verbs, each activating a different mode.
+
+To export a single existing resource, we'd use `export resource` as the verb. This would take the resource ID as a parameter and output the YAML file to stdout. The user would then be able to redirect the output to a file and modify it as necessary.
+
+For example, to export the YAML for an existing Virtual Network, the user would run:
+
+``` bash
+$ asoctl export resource http://management.azure.com/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/rg1/providers/Microsoft.Network/virtualNetworks/vnet1
+```
+
+As an extension, we could also support `export resourcegroup` to generate a single YAML file containing all the resources in a resource group. This would be useful for users who want to snapshot an existing resource group as YAML for reuse.
+
+``` bash
+$ asoctl export resourcegroup http://management.azure.com/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/rg1
```

Comment:
Rather than limit this to `resourcegroup` it feels like we should just call it `export-all` or something, and it works by exporting the resource in question and any subresources (as per Azure's ownership rules which we know).

That way if you wanted to for example export a Kubernetes cluster and all of its AgentPools, you just `export-all` the cluster. Or a MySQL Server, its firewall, and all of the firewall rules, you can just `export-all` the `MySQLServer`

Reply:
also maybe `export subscription` or something else to export all resources under a subscription?

Reply:
My initial instinct is that `export subscription` would be too wide a net, though it would be an easy extension if someone asked for it.

Reply:
Actually, I don't think we need a separate verb - we can make the decision based on what the URL points to.

Reply:
> Given ResourceGroup doesn't have a collection within, I thought it needed special treatment.

Not 100% sure I follow this? This is true for most resources which have subresources, there's no way to know if those subresources exist unless you:
1. Know about their types explicitly, and know the specific URL used to list them and call that.
2. You call a generic List Resources API with the resource as the prefix.

We know the full hierarchy of the types we support (that's the ownership stuff), _and_ we also have access to the generic List Resources API via the Go SDK. So either approach should allow us to treat ResourceGroups the same as other resources, I think?

Reply:
You may be right - I'll update the design to reflect this.